src/helpers/figures/charts/scorecard_chart_designer.ts

```diff
@@ -0,0 +1,377 @@
+import { Color } from "chart.js";
+import { DEFAULT_FONT } from "../../../constants";
+import { Figure, Pixel, PixelPosition, Style } from "../../../types";
+import { ScorecardChartRuntime } from "../../../types/chart";
+import { relativeLuminance } from "../../color";
+import { getFontSizeMatchingWidth } from "../../text_helper";
+
+/* Sizes of boxes containing the texts, in percentage of the Chart size */
+const TITLE_FONT_SIZE = 18;
+
+const BASELINE_BOX_HEIGHT_RATIO = 0.35;
+const KEY_BOX_HEIGHT_RATIO = 0.65;
+
+/** Baseline description should have a smaller font than the baseline */
+const BASELINE_DESCR_FONT_RATIO = 0.9;
+
+/* Padding at the border of the chart, in percentage of the chart width */
+const CHART_PADDING_RATIO = 0.02;
+
+/**
+ * Line height (in em)
+ * Having a line heigh =1em (=font size) don't work, the font will overflow.
+ */
+const LINE_HEIGHT = 1.2;
+
+type ScorecardChartDesignElement = {
+  text: string;
+  style: {
+    font: string;
+    color: Color;
+    strikethrough?: boolean;
+    underline?: boolean;
+  };
+  position: PixelPosition;
+};
+
+export type ScorecardChartDesign = {
+  canvas: {
+    width: number;
+    height: number;
+    backgroundColor: Color;
+  };
+  title?: ScorecardChartDesignElement;
+  baselineArrow?: ScorecardChartDesignElement;
+  baseline?: ScorecardChartDesignElement;
+  baselineDescr?: ScorecardChartDesignElement;
+  key?: ScorecardChartDesignElement;
+};
+
+export function formatBaselineDescr(
+  baselineDescr: string | undefined,
+  baseline: string | undefined
+): string {
+  const _baselineDescr = baselineDescr || "";
+  return baseline && _baselineDescr ? " " + _baselineDescr : _baselineDescr;
+}
+
+function getDefaultContextFont(
+  fontSize: number,
+  bold: boolean | undefined = false,
+  italic: boolean | undefined = false
+): string {
+  const italicStr = italic ? "italic" : "";
+  const weight = bold ? "bold" : "";
+  return `${italicStr} ${weight} ${fontSize}px ${DEFAULT_FONT}`;
+}
+
+export class ScorecardChartDesigner {
+  private context;
+  private width: number;
+  private height: number;
+
+  constructor(figure: Figure, readonly runtime: ScorecardChartRuntime) {
+    const canvas = document.createElement("canvas");
+    this.width = canvas.width = figure.width;
+    this.height = canvas.height = figure.height;
+    this.context = canvas.getContext("2d")!;
+  }
+
+  computeDesign(): ScorecardChartDesign {
```

Comment:
this method is way too long. Split it in smaller meaningful parts

Reply:
I give it a try, but each part need information coming from previous part for the positioning of new element. I tried to make it shorter (reduced from 120 lines to 80) and clearer.